File: server/api/test/file-security-policy.test.js
const assert = require('node:assert/strict');
const crypto = require('node:crypto');
const fs = require('node:fs/promises');
const os = require('node:os');
const path = require('node:path');
const test = require('node:test');
const JSZip = require('jszip');

const {
  FILE_SECURITY_MAX_FILES,
  FILE_SECURITY_MAX_NAME_LENGTH,
  FILE_SECURITY_ZIP_MAX_ENTRY_BYTES,
  FileSecurityPolicy,
  validateOfficeZipMetadata,
} = require('../src/modules/business-data/file-security-policy');
const {
  AttachmentUploadConfigService,
  sanitizeAttachmentOriginalName,
  validateTravelGroupAttachmentFile,
} = require('../src/modules/business-data/travel-group-attachment-storage.helper');

const DOCX_MIME =
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document';
const XLSX_MIME =
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';

test('FileSecurityPolicy validates harmless PDF, DOCX and XLSX fixtures', async () => {
  const fixtures = [
    {
      originalname: 'safe.pdf',
      mimetype: 'application/pdf',
      buffer: validPdf(),
    },
    {
      originalname: 'safe.docx',
      mimetype: DOCX_MIME,
      buffer: await openXml('docx'),
    },
    {
      originalname: 'safe.xlsx',
      mimetype: XLSX_MIME,
      buffer: await openXml('xlsx'),
    },
  ];
  for (const fixture of fixtures) {
    const validated = await validateTravelGroupAttachmentFile({
      ...fixture,
      size: fixture.buffer.length,
    });
    assert.equal(validated.originalName, fixture.originalname);
    assert.equal(validated.contentType, fixture.mimetype);
  }
});

test('FileSecurityPolicy rejects empty, mismatched, truncated and malformed documents', async () => {
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'empty.pdf',
      mimetype: 'application/pdf',
      buffer: Buffer.alloc(0),
      size: 0,
    }),
    'EMPTY_OR_INVALID_ATTACHMENT',
  );
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'renamed.pdf',
      mimetype: 'application/pdf',
      buffer: Buffer.from('not pdf'),
      size: 7,
    }),
    'ATTACHMENT_CONTENT_MISMATCH',
  );
  const validDocx = await openXml('docx');
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'wrong.docx',
      mimetype: 'application/pdf',
      buffer: validDocx,
      size: validDocx.length,
    }),
    'UNSUPPORTED_ATTACHMENT_TYPE',
  );
  const truncatedPdf = Buffer.from('%PDF-1.4\n1 0 obj\n<<>>\nendobj\n');
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'truncated.pdf',
      mimetype: 'application/pdf',
      buffer: truncatedPdf,
      size: truncatedPdf.length,
    }),
    'TRUNCATED_PDF',
  );
  const damagedZip = Buffer.from([
    0x50, 0x4b, 0x03, 0x04, ...Buffer.alloc(40),
  ]);
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'damaged.docx',
      mimetype: DOCX_MIME,
      buffer: damagedZip,
      size: damagedZip.length,
    }),
    'INVALID_OFFICE_ZIP',
  );
  const missingStructure = await new JSZip()
    .file('[Content_Types].xml', '<Types />')
    .generateAsync({ type: 'nodebuffer' });
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'missing.xlsx',
      mimetype: XLSX_MIME,
      buffer: missingStructure,
      size: missingStructure.length,
    }),
    'INVALID_OFFICE_STRUCTURE',
  );
});

test('FileSecurityPolicy rejects unsafe names, macro formats and active content', async () => {
  assert.equal(
    sanitizeAttachmentOriginalName('../../folder\\safe.pdf'),
    'safe.pdf',
  );
  assert.throws(
    () => FileSecurityPolicy.assertSafeOriginalName('../unsafe.pdf'),
    (error) => error?.code === 'INVALID_ATTACHMENT_FILE_NAME',
  );
  assert.throws(
    () =>
      FileSecurityPolicy.assertSafeOriginalName(
        `${'a'.repeat(FILE_SECURITY_MAX_NAME_LENGTH)}.pdf`,
      ),
    (error) => error?.code === 'INVALID_ATTACHMENT_FILE_NAME',
  );
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'macro.docm',
      mimetype: DOCX_MIME,
      buffer: await openXml('docx'),
      size: 100,
    }),
    'MACRO_ENABLED_DOCUMENT_REJECTED',
  );
  const activePdf = Buffer.from(
    '%PDF-1.4\n1 0 obj\n<< /JavaScript 2 0 R >>\nendobj\n' +
      'trailer\n<<>>\nstartxref\n9\n%%EOF\n',
  );
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'active.pdf',
      mimetype: 'application/pdf',
      buffer: activePdf,
      size: activePdf.length,
    }),
    'ACTIVE_PDF_CONTENT_REJECTED',
  );
});

test('ZIP metadata limits use simulated sizes and reject traversal, nesting and duplicates', () => {
  assert.equal(FILE_SECURITY_MAX_FILES, 5);
  assert.throws(
    () =>
      validateOfficeZipMetadata([
        metadataEntry('word/document.xml', {
          compressedSize: 1,
          uncompressedSize: 101,
        }),
      ]),
    (error) => error?.code === 'ZIP_COMPRESSION_RATIO_EXCEEDED',
  );
  assert.throws(
    () =>
      validateOfficeZipMetadata([
        metadataEntry('xl/workbook.xml', {
          uncompressedSize: FILE_SECURITY_ZIP_MAX_ENTRY_BYTES + 1,
        }),
      ]),
    (error) => error?.code === 'ZIP_SIZE_LIMIT_EXCEEDED',
  );
  assert.throws(
    () => validateOfficeZipMetadata([metadataEntry('../escape.xml')]),
    (error) => error?.code === 'ZIP_PATH_REJECTED',
  );
  assert.throws(
    () =>
      validateOfficeZipMetadata([
        metadataEntry('word/document.xml'),
        metadataEntry('word/document.xml'),
      ]),
    (error) => error?.code === 'DUPLICATE_ZIP_ENTRY_REJECTED',
  );
  assert.throws(
    () =>
      validateOfficeZipMetadata([
        metadataEntry('word/embeddings/nested.zip'),
      ]),
    (error) => error?.code === 'EMBEDDED_OFFICE_CONTENT_REJECTED',
  );
  assert.throws(
    () =>
      validateOfficeZipMetadata([
        {
          ...metadataEntry('word/document.xml'),
          isEncrypted: () => true,
        },
      ]),
    (error) => error?.code === 'ENCRYPTED_ZIP_REJECTED',
  );
});

test('OOXML validation rejects external relationships and embedded objects', async () => {
  const linked = await openXml('docx', {
    extraEntries: {
      'word/_rels/document.xml.rels':
        '<Relationships><Relationship TargetMode="External" ' +
        'Target="https://invalid.example/" /></Relationships>',
    },
  });
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'linked.docx',
      mimetype: DOCX_MIME,
      buffer: linked,
      size: linked.length,
    }),
    'EXTERNAL_OFFICE_RELATIONSHIP_REJECTED',
  );
  const embedded = await openXml('xlsx', {
    extraEntries: {
      'xl/embeddings/oleObject1.bin': Buffer.from('harmless-placeholder'),
    },
  });
  await rejectsCode(
    validateTravelGroupAttachmentFile({
      originalname: 'embedded.xlsx',
      mimetype: XLSX_MIME,
      buffer: embedded,
      size: embedded.length,
    }),
    'EMBEDDED_OFFICE_CONTENT_REJECTED',
  );
});

test('startup cleanup removes only expired randomized upload files', async () => {
  const root = await fs.mkdtemp(
    path.join(os.tmpdir(), 'jiangjiu-file-policy-'),
  );
  const temporaryRoot = path.join(root, 'temporary');
  const permanentRoot = path.join(root, 'permanent');
  const previousTemp = process.env.ATTACHMENT_UPLOAD_TEMP_DIR;
  const previousPermanent = process.env.TRAVEL_GROUP_ATTACHMENT_DIR;
  process.env.ATTACHMENT_UPLOAD_TEMP_DIR = temporaryRoot;
  process.env.TRAVEL_GROUP_ATTACHMENT_DIR = permanentRoot;
  try {
    await fs.mkdir(temporaryRoot, { recursive: true });
    const expired = path.join(
      temporaryRoot,
      `.upload-${crypto.randomUUID()}`,
    );
    const current = path.join(
      temporaryRoot,
      `.upload-${crypto.randomUUID()}`,
    );
    const unrelated = path.join(temporaryRoot, 'keep.txt');
    await fs.writeFile(expired, 'expired');
    await fs.writeFile(current, 'current');
    await fs.writeFile(unrelated, 'unrelated');
    const old = new Date(Date.now() - 48 * 60 * 60 * 1000);
    await fs.utimes(expired, old, old);

    const config = new AttachmentUploadConfigService();
    await config.onModuleInit();

    assert.equal(await exists(expired), false);
    assert.equal(await exists(current), true);
    assert.equal(await exists(unrelated), true);
  } finally {
    restoreEnv('ATTACHMENT_UPLOAD_TEMP_DIR', previousTemp);
    restoreEnv('TRAVEL_GROUP_ATTACHMENT_DIR', previousPermanent);
    await fs.rm(root, { recursive: true, force: true });
  }
});

function validPdf() {
  return Buffer.from(
    '%PDF-1.4\n1 0 obj\n<< /Type /Catalog >>\nendobj\n' +
      'trailer\n<< /Root 1 0 R >>\nstartxref\n9\n%%EOF\n',
  );
}

async function openXml(kind, { extraEntries = {} } = {}) {
  const zip = new JSZip();
  zip.file(
    '[Content_Types].xml',
    kind === 'docx'
      ? '<Types><Override PartName="/word/document.xml" /></Types>'
      : '<Types><Override PartName="/xl/workbook.xml" /></Types>',
  );
  zip.file(
    kind === 'docx' ? 'word/document.xml' : 'xl/workbook.xml',
    kind === 'docx' ? '<document />' : '<workbook />',
  );
  for (const [name, content] of Object.entries(extraEntries)) {
    zip.file(name, content);
  }
  return zip.generateAsync({
    type: 'nodebuffer',
    compression: 'DEFLATE',
  });
}

function metadataEntry(
  fileNameStr,
  { compressedSize = 10, uncompressedSize = 10 } = {},
) {
  return {
    fileNameStr,
    compressedSize,
    uncompressedSize,
    dir: false,
    unixPermissions: 0,
    isEncrypted: () => false,
  };
}

async function rejectsCode(promise, code) {
  await assert.rejects(promise, (error) => error?.code === code);
}

async function exists(value) {
  try {
    await fs.stat(value);
    return true;
  } catch (error) {
    if (error?.code === 'ENOENT') return false;
    throw error;
  }
}

function restoreEnv(name, value) {
  if (value === undefined) {
    delete process.env[name];
  } else {
    process.env[name] = value;
  }
}

File: apps/mobile_desktop/test/file_security_policy_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:archive/archive.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:jiangjiu_mobile_desktop/core/file_security_policy.dart';
import 'package:path/path.dart' as path;

void main() {
  late Directory testRoot;

  setUp(() async {
    testRoot = await Directory.systemTemp.createTemp(
      'jiangjiu-file-security-',
    );
    FileSecurityPolicy.temporaryRootProviderForTesting = () async => testRoot;
  });

  tearDown(() async {
    FileSecurityPolicy.temporaryRootProviderForTesting = null;
    if (await testRoot.exists()) {
      await testRoot.delete(recursive: true);
    }
  });

  test('accepts harmless PDF, DOCX and XLSX into random private temp names',
      () async {
    final fixtures = <({String name, String mime, Uint8List bytes})>[
      (
        name: 'safe.pdf',
        mime: 'application/pdf',
        bytes: Uint8List.fromList(_validPdf()),
      ),
      (
        name: 'safe.docx',
        mime:
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        bytes: _openXml(wordDocument: true),
      ),
      (
        name: 'safe.xlsx',
        mime:
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        bytes: _openXml(wordDocument: false),
      ),
    ];

    for (final fixture in fixtures) {
      final prepared = await FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(fixture.name, fixture.mime, fixture.bytes),
        ],
      );
      expect(prepared.single.fileName, fixture.name);
      expect(
        path.basename(prepared.single.path),
        matches(RegExp(r'^\.upload-[0-9a-f]{32}$')),
      );
      expect(await File(prepared.single.path).readAsBytes(), fixture.bytes);
      await FileSecurityPolicy.cleanupPreparedUploads(prepared);
      expect(await File(prepared.single.path).exists(), isFalse);
    }
  });

  test('rejects empty, mismatched MIME, renamed and truncated documents',
      () async {
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source('empty.pdf', 'application/pdf', Uint8List(0)),
        ],
      ),
      'EMPTY_FILE',
    );
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'wrong.pdf',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            Uint8List.fromList(_validPdf()),
          ),
        ],
      ),
      'MIME_CONTENT_MISMATCH',
    );
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'renamed.pdf',
            'application/pdf',
            Uint8List.fromList(utf8.encode('not a PDF')),
          ),
        ],
      ),
      'FILE_CONTENT_MISMATCH',
    );
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'truncated.pdf',
            'application/pdf',
            Uint8List.fromList(
              ascii.encode('%PDF-1.4\n1 0 obj\n<<>>\nendobj\n'),
            ),
          ),
        ],
      ),
      'TRUNCATED_PDF',
    );
  });

  test('rejects damaged or structurally invalid Office ZIP containers',
      () async {
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'damaged.docx',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            Uint8List.fromList(<int>[
              0x50,
              0x4b,
              0x03,
              0x04,
              ...List<int>.filled(40, 0),
            ]),
          ),
        ],
      ),
      'ZIP_ENTRY_LIMIT_EXCEEDED',
    );
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'missing.xlsx',
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
            _zip(<String, List<int>>{
              '[Content_Types].xml': utf8.encode('<Types />'),
            }),
          ),
        ],
      ),
      'INVALID_OFFICE_STRUCTURE',
    );
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'traversal.docx',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            _openXml(
              wordDocument: true,
              extraEntries: <String, List<int>>{
                '../escape.xml': utf8.encode('<safe />'),
              },
            ),
          ),
        ],
      ),
      'ZIP_PATH_REJECTED',
    );
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'nested.xlsx',
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
            _openXml(
              wordDocument: false,
              extraEntries: <String, List<int>>{
                'xl/embeddings/nested.zip': const <int>[0x50, 0x4b],
              },
            ),
          ),
        ],
      ),
      'EMBEDDED_OFFICE_CONTENT_REJECTED',
    );
  });

  test('sanitizes names and rejects macros, count, single and total limits',
      () async {
    expect(
      FileSecurityPolicy.sanitizeFileName(
        '../../folder\\bad\r\nname.pdf',
      ),
      'bad__name.pdf',
    );
    expect(
      FileSecurityPolicy.sanitizeFileName(
        '${List<String>.filled(240, 'a').join()}.pdf',
      ).length,
      fileSecurityMaxNameLength,
    );
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          _source(
            'macro.docm',
            'application/vnd.ms-word.document.macroEnabled.12',
            _openXml(wordDocument: true),
          ),
        ],
      ),
      'MACRO_ENABLED_DOCUMENT_REJECTED',
    );
    expect(
      () => FileSecurityPolicy.validateAggregateMetadata(
        List<int>.filled(fileSecurityMaxFiles + 1, 1),
      ),
      throwsA(_violationCode('FILE_COUNT_LIMIT')),
    );
    expect(
      () => FileSecurityPolicy.validateAggregateMetadata(
        <int>[fileSecurityMaxFileBytes + 1],
      ),
      throwsA(_violationCode('FILE_TOO_LARGE')),
    );
    expect(
      () => FileSecurityPolicy.validateAggregateMetadata(
        const <int>[
          9 * 1024 * 1024,
          9 * 1024 * 1024,
          8 * 1024 * 1024,
        ],
      ),
      throwsA(_violationCode('TOTAL_SIZE_LIMIT')),
    );
  });

  test('stream failures and orphan cleanup do not leave private temp files',
      () async {
    await _expectCode(
      FileSecurityPolicy.prepareUploads(
        <FileSecuritySource>[
          FileSecuritySource(
            fileName: 'cancelled.pdf',
            declaredMimeType: 'application/pdf',
            openRead: () => Stream<List<int>>.multi((controller) {
              controller.add(ascii.encode('%PDF-1.4\n'));
              controller.addError(StateError('simulated cancellation'));
            }),
          ),
        ],
      ),
      null,
      errorType: StateError,
    );
    final privateDirectory = Directory(
      path.join(testRoot.path, 'jiangjiu-secure-uploads'),
    );
    expect(await privateDirectory.list().toList(), isEmpty);

    final expired = File(
      path.join(
        privateDirectory.path,
        '.upload-${List<String>.filled(32, 'a').join()}',
      ),
    );
    final current = File(
      path.join(
        privateDirectory.path,
        '.upload-${List<String>.filled(32, 'b').join()}',
      ),
    );
    final unrelated = File(path.join(privateDirectory.path, 'keep.txt'));
    await expired.writeAsString('expired');
    await current.writeAsString('current');
    await unrelated.writeAsString('unrelated');
    await expired.setLastModified(DateTime(2026, 1, 1));

    await FileSecurityPolicy.cleanupOrphanedTemporaryFiles(
      now: DateTime(2026, 1, 3),
    );
    expect(await expired.exists(), isFalse);
    expect(await current.exists(), isTrue);
    expect(await unrelated.exists(), isTrue);
  });
}

FileSecuritySource _source(String name, String mime, Uint8List bytes) {
  return FileSecuritySource(
    fileName: name,
    declaredMimeType: mime,
    openRead: () => Stream<List<int>>.value(bytes),
  );
}

List<int> _validPdf() {
  return ascii.encode(
    '%PDF-1.4\n1 0 obj\n<< /Type /Catalog >>\nendobj\n'
    'trailer\n<< /Root 1 0 R >>\nstartxref\n9\n%%EOF\n',
  );
}

Uint8List _openXml({
  required bool wordDocument,
  Map<String, List<int>> extraEntries = const <String, List<int>>{},
}) {
  return _zip(<String, List<int>>{
    '[Content_Types].xml': utf8.encode(
      wordDocument
          ? '<Types><Override PartName="/word/document.xml" /></Types>'
          : '<Types><Override PartName="/xl/workbook.xml" /></Types>',
    ),
    wordDocument ? 'word/document.xml' : 'xl/workbook.xml':
        utf8.encode(wordDocument ? '<document />' : '<workbook />'),
    ...extraEntries,
  });
}

Uint8List _zip(Map<String, List<int>> entries) {
  final archive = Archive();
  for (final entry in entries.entries) {
    archive.addFile(ArchiveFile.bytes(entry.key, entry.value));
  }
  return Uint8List.fromList(ZipEncoder().encode(archive));
}

Future<void> _expectCode(
  Future<Object?> future,
  String? code, {
  Type? errorType,
}) async {
  await expectLater(
    future,
    throwsA(
      predicate<Object>(
        (error) =>
            (code == null ||
                error is FileSecurityViolation && error.code == code) &&
            (errorType == null || error.runtimeType == errorType),
      ),
    ),
  );
}

Matcher _violationCode(String code) {
  return isA<FileSecurityViolation>().having(
    (error) => error.code,
    'code',
    code,
  );
}
